<a href="https://colab.research.google.com/github/surajnik77/ML-Algorithm/blob/main/Naive_Bayes_tips_dataset.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder,OneHotEncoder,OrdinalEncoder
from sklearn.naive_bayes import GaussianNB,MultinomialNB,BernoulliNB
from sklearn.metrics import confusion_matrix,classification_report

In [2]:
df = sns.load_dataset("tips")
df.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype   
---  ------      --------------  -----   
 0   total_bill  244 non-null    float64 
 1   tip         244 non-null    float64 
 2   sex         244 non-null    category
 3   smoker      244 non-null    category
 4   day         244 non-null    category
 5   time        244 non-null    category
 6   size        244 non-null    int64   
dtypes: category(4), float64(2), int64(1)
memory usage: 7.4 KB


In [4]:
df.describe()

total_bill         tip        size
count  244.000000  244.000000  244.000000
mean    19.785943    2.998279    2.569672
std      8.902412    1.383638    0.951100
min      3.070000    1.000000    1.000000
25%     13.347500    2.000000    2.000000
50%     17.795000    2.900000    2.000000
75%     24.127500    3.562500    3.000000
max     50.810000   10.000000    6.000000

In [5]:
df.isna().sum()

total_bill    0
tip           0
sex           0
smoker        0
day           0
time          0
size          0
dtype: int64

In [6]:
df.shape

(244, 7)

In [7]:
df.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

**we have here categorical data in features so encoded this data.**

#Data Preprocessing

# here we have to predict time of dinner on the basis of other features

In [8]:
df.time.unique()

['Dinner', 'Lunch']
Categories (2, object): ['Lunch', 'Dinner']

In [9]:
# OneHotEncoder-  Encoding the one column categorical data into seperart categorical column.
onehotencoder = OneHotEncoder(handle_unknown='ignore')
onehotencoder.fit(df[['sex']])
encoded = onehotencoder.transform(df[['sex']])
df[onehotencoder.categories_[0]] = encoded.toarray()

In [10]:
df.head()

total_bill   tip     sex smoker  day    time  size  Female  Male
0       16.99  1.01  Female     No  Sun  Dinner     2     1.0   0.0
1       10.34  1.66    Male     No  Sun  Dinner     3     0.0   1.0
2       21.01  3.50    Male     No  Sun  Dinner     3     0.0   1.0
3       23.68  3.31    Male     No  Sun  Dinner     2     0.0   1.0
4       24.59  3.61  Female     No  Sun  Dinner     4     1.0   0.0

In [11]:
#OrdinalEncoder - Encoding Data in order or rank form
oe = OrdinalEncoder()
df[['smoker','day']] = oe.fit_transform(df[['smoker','day']])
df.head()


total_bill   tip     sex  smoker  day    time  size  Female  Male
0       16.99  1.01  Female     0.0  2.0  Dinner     2     1.0   0.0
1       10.34  1.66    Male     0.0  2.0  Dinner     3     0.0   1.0
2       21.01  3.50    Male     0.0  2.0  Dinner     3     0.0   1.0
3       23.68  3.31    Male     0.0  2.0  Dinner     2     0.0   1.0
4       24.59  3.61  Female     0.0  2.0  Dinner     4     1.0   0.0

In [12]:
oe.categories_

[array(['No', 'Yes'], dtype=object),
 array(['Fri', 'Sat', 'Sun', 'Thur'], dtype=object)]

In [13]:
X = df.drop(['sex','time'],axis=1)
X[:5]

total_bill   tip  smoker  day  size  Female  Male
0       16.99  1.01     0.0  2.0     2     1.0   0.0
1       10.34  1.66     0.0  2.0     3     0.0   1.0
2       21.01  3.50     0.0  2.0     3     0.0   1.0
3       23.68  3.31     0.0  2.0     2     0.0   1.0
4       24.59  3.61     0.0  2.0     4     1.0   0.0

In [14]:
y = df.time
y[:5]

0    Dinner
1    Dinner
2    Dinner
3    Dinner
4    Dinner
Name: time, dtype: category
Categories (2, object): ['Lunch', 'Dinner']

In [15]:
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.20,random_state=42)

In [16]:
le = LabelEncoder()
le.fit(y_train)

LabelEncoder()

In [17]:
le.classes_

array(['Dinner', 'Lunch'], dtype=object)

In [18]:
y_train = le.transform(y_train)
y_test = le.transform(y_test)

In [19]:
X_train[:5]

total_bill   tip  smoker  day  size  Female  Male
228       13.28  2.72     0.0  1.0     2     0.0   1.0
208       24.27  2.03     1.0  1.0     2     0.0   1.0
96        27.28  4.00     1.0  0.0     2     0.0   1.0
167       31.71  4.50     0.0  2.0     4     0.0   1.0
84        15.98  2.03     0.0  3.0     2     0.0   1.0

In [20]:
y_train[:5]

array([0, 0, 0, 0, 1])

In [21]:
#GaussianNB
gnb = GaussianNB()
gnb.fit(X_train,y_train)

GaussianNB()

In [22]:
y_pred = gnb.predict(X_test)
y_pred[:5]

array([0, 0, 0, 0, 1])

In [23]:
y_test[:5]

array([0, 0, 0, 0, 1])

In [24]:
print(confusion_matrix(y_pred,y_test))
print(classification_report(y_pred,y_test))

[[32  2]
 [ 2 13]]
              precision    recall  f1-score   support

           0       0.94      0.94      0.94        34
           1       0.87      0.87      0.87        15

    accuracy                           0.92        49
   macro avg       0.90      0.90      0.90        49
weighted avg       0.92      0.92      0.92        49



In [25]:
bnb = BernoulliNB()
bnb.fit(X_train,y_train)

BernoulliNB()

In [26]:
y_pred1 = bnb.predict(X_test)
y_pred1[:5]

array([0, 0, 0, 0, 0])

In [27]:
print(confusion_matrix(y_pred1,y_test))
print(classification_report(y_pred1,y_test))

[[33 15]
 [ 1  0]]
              precision    recall  f1-score   support

           0       0.97      0.69      0.80        48
           1       0.00      0.00      0.00         1

    accuracy                           0.67        49
   macro avg       0.49      0.34      0.40        49
weighted avg       0.95      0.67      0.79        49



In [28]:
mnb = MultinomialNB()
mnb.fit(X_train,y_train)

MultinomialNB()

In [29]:
y_pred2 = mnb.predict(X_test)
y_pred2[:5]

array([0, 0, 0, 0, 1])

In [30]:
print(confusion_matrix(y_pred2,y_test))
print(classification_report(y_pred2,y_test))

[[32  5]
 [ 2 10]]
              precision    recall  f1-score   support

           0       0.94      0.86      0.90        37
           1       0.67      0.83      0.74        12

    accuracy                           0.86        49
   macro avg       0.80      0.85      0.82        49
weighted avg       0.87      0.86      0.86        49



**here we select GaussianNB Model for final modeling with Accuracy 92%**